internal/stores/hostdb.go

```diff
@@ -181,3 +195,236 @@ func NewJSONHostDB(dir string) (*JSONHostDB, modules.ConsensusChangeID, error) {
+
+	return &SQLHostDB{
+		staticDB: db,
+	}, ccid, nil
+}
+
+// Host returns information about a host.
+func (db *SQLHostDB) Host(hostKey consensus.PublicKey) (hostdb.Host, error) {
+	var h host
+	tx := db.staticDB.Where(&host{PublicKey: hostKey[:]}).
+		Preload("Interactions").
+		Preload("Announcements").
+		Find(&h)
+	if tx.Error == nil && tx.RowsAffected == 0 {
+		return hostdb.Host{}, ErrHostNotFound
+	}
+	return h.Host(), tx.Error
+}
+
+// Hosts returns up to max hosts that have not been interacted with since
+// the specified time.
+func (db *SQLHostDB) Hosts(notSince time.Time, max int) ([]hostdb.Host, error) {
+	lastInteraction := func(h hostdb.Host) time.Time {
+		if len(h.Interactions) == 0 {
+			return time.Time{}
+		}
+		return h.Interactions[len(h.Interactions)-1].Timestamp
+	}
+	hosts, err := db.SelectHosts(math.MaxInt, func(h hostdb.Host) bool { return true })
+	if err != nil {
+		return nil, err
+	}
+	for _, host := range hosts {
+		if len(hosts) == max {
+			break
+		} else if lastInteraction(host).Before(notSince) {
+			hosts = append(hosts, host)
+		}
+	}
+	return hosts, nil
+}
+
+// RecordInteraction records an interaction with a host. If the host is not in
+// the store, a new entry is created for it.
+func (db *SQLHostDB) RecordInteraction(hostKey consensus.PublicKey, hi hostdb.Interaction) error {
+	return db.staticDB.Transaction(func(tx *gorm.DB) error {
+		// Create a host if it doesn't exist yet.
+		if err := tx.FirstOrCreate(&host{}, &host{PublicKey: hostKey[:]}).Error; err != nil {
+			return err
+		}
+
+		// Create an interaction.
+		return tx.Create(&interaction{
+			Host:      hostKey[:],
+			Timestamp: hi.Timestamp.UTC(), // use UTC in DB
+			Type:      hi.Type,
+			Result:    hi.Result,
+		}).Error
+	})
+}
+
+// SelectHosts returns up to n hosts for which the supplied filter returns true.
+func (db *SQLHostDB) SelectHosts(n int, filter func(hostdb.Host) bool) ([]hostdb.Host, error) {
+	var hosts []host
+	tx := db.staticDB.Preload("Interactions").
+		Preload("Announcements").
+		Find(&hosts)
+	if tx.Error != nil {
+		return nil, tx.Error
+	}
+
+	var drawnHosts []hostdb.Host
+	for i := 0; i < len(hosts) && len(drawnHosts) < n; i++ {
+		h := hosts[i].Host()
+		if filter(h) {
+			drawnHosts = append(drawnHosts, hosts[i].Host())
+		}
+	}
+	return drawnHosts, nil
+}
+
+// ProcessConsensusChange implements consensus.Subscriber.
+func (db *SQLHostDB) ProcessConsensusChange(cc modules.ConsensusChange) {
+	height := cc.InitialHeight()
+	for range cc.RevertedBlocks {
+		height--
+	}
+
+	// Atomically apply ConsensusChange.
+	err := db.staticDB.Transaction(func(tx *gorm.DB) error {
+		for _, b := range cc.AppliedBlocks {
+			hostdb.ForEachAnnouncement(b, height, func(hostKey consensus.PublicKey, ha hostdb.Announcement) error {
+				return insertAnnouncement(tx, hostKey, ha)
+			})
+			height++
+		}
+		return tx.Model(&consensusInfo{}).Where(&consensusInfo{ID: consensusInfoID}).Update("CCID", cc.ID[:]).Error
+	})
```

Comment:
you can use a closure around a sticky error here; that way, you don't need to change the signature of `ForEachAnnouncement`:
```go
// Atomically apply ConsensusChange.
err := db.staticDB.Transaction(func(tx *gorm.DB) error {
	var err error
	for _, b := range cc.AppliedBlocks {
		hostdb.ForEachAnnouncement(b, height, func(hostKey consensus.PublicKey, ha hostdb.Announcement) {
	 		if err == nil {
			 	err = insertAnnouncement(tx, hostKey, ha)
	 		}
		})
		height++
	}
	if err != nil {
		return err
	}
	return tx.Model(&consensusInfo{}).Where(&consensusInfo{ID: consensusInfoID}).Update("CCID", cc.ID[:]).Error
})
```

Reply:
Done. I thought it might be nice to abort `ForEachAnnouncement` but I guess it's not that big of a deal.